store/src/java/com/zimbra/cs/filter/SieveVisitor.java

```diff
@@ -449,16 +459,28 @@ private void acceptTest(Node node, RuleProperties props) throws ServiceException
                 headers = getMultiValue(node, 0, nextArgIndex, 0);
                 value = getValue(node, 0, nextArgIndex + 1, 0, 0);
                 validateCountComparator(isCount, comparator);
-                if (valueComparison != null) { 
-                    visitAddressTest(node, VisitPhase.begin, props, headers, part, valueComparison, isCount, value);
-                    accept(node, props);
-                    visitAddressTest(node, VisitPhase.end, props, headers, part, valueComparison, isCount, value);
-                } else {
-                    visitAddressTest(node, VisitPhase.begin, props, headers, part, comparison, caseSensitive, value);
-                    accept(node, props);
-                    visitAddressTest(node, VisitPhase.end, props, headers, part, comparison, caseSensitive, value);
-                }
 
+                if ("address".equalsIgnoreCase(nodeName)) {
+                    if (valueComparison != null) {
+                        visitAddressTest(node, VisitPhase.begin, props, headers, part, valueComparison, isCount, value);
+                        accept(node, props);
+                        visitAddressTest(node, VisitPhase.end, props, headers, part, valueComparison, isCount, value);
+                    } else {
+                        visitAddressTest(node, VisitPhase.begin, props, headers, part, comparison, caseSensitive, value);
+                        accept(node, props);
+                        visitAddressTest(node, VisitPhase.end, props, headers, part, comparison, caseSensitive, value);
+                    }
+                } else if ("envelope".equalsIgnoreCase(nodeName)) {
```

Comment:
Why don't you use "else { ... } in this context?  In general, "else if { ... }" expects the "else { ... }", and it looks as if there might be another option of the nodeName than "address" and "envelope" which is none in this block.